### PR TITLE
Add blocksparse example and while lowering

### PR DESCRIPTION
**Pull request by sharadmv**

### CODE REVIEW

File: examples/pallas/blocksparse_matmul.py [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+    block_indices = jnp.sort(random.permutation(k, jnp.arange(num_cols))[:max_num_blocks])
+    for j, b in zip(range(max_num_blocks), block_indices):
+      if j < num_blocks_in_row:
+        index = [b, block_index]
+        block_index += 1
+      else:
+        index = [0, 0]
+      row.append(index)
+    indices.append(row)
+  indices = jnp.array(indices)
```
Comment (jakevdp) [inline examples/pallas/blocksparse_matmul.py:157]: Am I reading this right that `indices` is three dimensions? How does this square with the index shape being `[n_rows, n_blocks]` above?

**Review by jakevdp — COMMENTED** [review]
Looks great!

File: examples/pallas/blocksparse_matmul.py (hunk 1 above)
Comment (sharadmv) [reply examples/pallas/blocksparse_matmul.py:157]: Oops that comment is wrong. `indices` is `int32[n_rows, max_num_blocks_per_row, 2]`

File: examples/pallas/blocksparse_matmul.py [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+  def __matmul__(self, other):
+    assert isinstance(other, jnp.ndarray)
+    self._validate()
+    assert self.ndim == other.ndim == 2
+    assert self.shape[1] == other.shape[0]
+    out = jnp.zeros((self.shape[0], other.shape[1]),
+                    dtype=jnp.result_type(self.dtype, other.dtype))
+    _, n, m = self.blocks.shape
+    nrows = self.shape[0] // n
+    def i_body(i, out):
```
[the remark this answers is not in the document]
Comment (sharadmv) [reply examples/pallas/blocksparse_matmul.py:122]: Done! It seems to have sped up the example by about 10x.

**Review by zhangqiaorjc — APPROVED** [review]

**Pull request merged**